audio/mods/impulsetracker.cpp

```diff
@@ -0,0 +1,215 @@
+		reader->stream = stream;
+	}
+	return (MREADER *)reader;
+}
+
+static BOOL memoryReaderEof(MREADER *reader) {
+	MikMemoryReader *mr = (MikMemoryReader *)reader;
+	if (!mr)
+		return 1;
+	if (mr->stream && mr->stream->eos() == true)
+		return 1;
+	return 0;
+}
+
+static BOOL memoryReaderRead(MREADER *reader, void *ptr, size_t size) {
+	MikMemoryReader *mr;
+	mr = (MikMemoryReader *)reader;
+	if (!mr && !mr->stream)
+		return 0;
+
+	uint32 receivedBytes = mr->stream->read(ptr, size);
+	if (receivedBytes < size)
+		return 0; // not enough remaining bytes (or error)
+	return 1;
+}
+
+static int memoryReaderGet(MREADER *reader) {
+	MikMemoryReader *mr;
+	mr = (MikMemoryReader *)reader;
+	if (!mr->stream)
+		return -1;
+	return mr->stream->readByte();
+}
+
+static int memoryReaderSeek(MREADER *reader, long offset, int whence) {
+	MikMemoryReader *mr;
+	mr = (MikMemoryReader *)reader;
+	if (!reader || !mr->stream)
+		return -1;
+
+	return mr->stream->seek(offset, whence);
+}
+
+static long memoryReaderTell(MREADER *reader) {
+	if (reader)
+		return ((MikMemoryReader *)reader)->stream->pos();
+	return 0;
+}
+// End memory wrappper
+
+namespace Audio {
+class ImpulseTrackerMod : public RewindableAudioStream {
+public:
+	ImpulseTrackerMod(Common::SeekableReadStream *stream, DisposeAfterUse::Flag disposeAfterUse);
+	~ImpulseTrackerMod();
+
+	// ImpulseTrackerMod functions
+	bool isLoaded() {
+		return _mikmod_load_successful;
+	}
+
+	// AudioStream API
+	int readBuffer(int16 *buffer, const int numSamples) override {
+		// Multiplied by 2 as VC_WriteBytes function expects 8 byte integer arrays, whereas buffer needs 16 ones.
+		VC_WriteBytes((SBYTE *)buffer, numSamples * 2);
+		return numSamples;
+	}
+
+	bool isStereo() const override {
+		return true;
+	}
+	int getRate() const override {
+		return 44100;
+	}
+	bool endOfData() const override {
+		return !Player_Active();
+	}
+	bool endOfStream() const override {
+		return endOfData();
+	}
+
+	// RewindableAudioStream API
+	bool rewind() {
+		Player_SetPosition(0);
+		return true;
+	}
+
+private:
+	DisposeAfterUse::Flag _dispose;
+	bool _mikmod_load_successful = false;
+	Common::SeekableReadStream *_stream;
+	MREADER *_reader;
+	MODULE *_mod;
+};
+
+ImpulseTrackerMod::ImpulseTrackerMod(Common::SeekableReadStream *stream, DisposeAfterUse::Flag disposeAfterUse) {
+	if (!stream) {
+		warning("ImpulseTrackerMod::ImpulseTrackerMod(): Input file/stream is invalid.");
+		return;
+	}
+
+	MikMod_InitThreads();
+
+	// No sound driver as we will be routing through scrumm mixer!
+	MikMod_RegisterDriver(&drv_nos);
+
+	// Set flags
+	md_mode |= DMODE_SOFT_MUSIC | DMODE_NOISEREDUCTION;
+	md_mixfreq = getRate();
+	if (MikMod_Init("")) {
+		warning("ImpulseTrackerMod::ImpulseTrackerMod(): Could not initialize sound, reason: %s\n",
+		        MikMod_strerror(MikMod_errno));
+		return;
+	}
+
+	// Loading only impulse tracker loader!
+	MikMod_RegisterLoader(&load_it);
+
+
+	// TODO: Not currently in use
```

Comment:
Is it truly not in use now?

Reply:
It is, however wrapped in MREADER instance, this is like for now we just remove this class parameter __stream_ and still the program will work